File: hustack/Ex16_LIS/codes/segtree.py
class SegmentTree:
    def __init__(self, n):
        self.N = 1
        while self.N < n: self.N <<= 1
        self.tree = [0] * (2 * self.N)

    def update(self, i, v):
        i += self.N
        self.tree[i] = max(self.tree[i], v)
        i >>= 1
        while i:
            self.tree[i] = max(self.tree[2*i], self.tree[2*i+1])
            i >>= 1

    def query(self, l, r):
        l += self.N; r += self.N
        res = 0
        while l <= r:
            if (l & 1) == 1:
                res = max(res, self.tree[l])
                l += 1
            if (r & 1) == 0:
                res = max(res, self.tree[r])
                r -= 1
            l >>= 1; r >>= 1
        return res
    
if __name__ == "__main__":
    n = int(input())
    a = list(map(int, input().split()))

    vals = sorted(set(a))
    comp = {v:i for i,v in enumerate(vals)}
    st = SegmentTree(len(vals))
    best = 0
    for x in a:
        idx = comp[x]
        if idx > 0:
            cur = st.query(0, idx-1) + 1
        else:
            cur = 1
        st.update(idx, cur)
        best = max(best, cur)
        
    print(best)

File: hustack/Ex16_LIS/codes/segtree.cpp
#include <bits/stdc++.h>
using namespace std;

struct SegTree {
    int n;
    vector<int> t;
    SegTree(int _n): n(1) {
        while (n < _n) n <<= 1;
        t.assign(2 * n, 0);
    }
    void update(int i, int v) {
        i += n;
        t[i] = max(t[i], v);
        for (i >>= 1; i; i >>= 1)
            t[i] = max(t[2 * i], t[2 * i+1]);
    }
    int query(int l, int r) {
        int res = 0;
        for (l += n, r += n; l <= r; l >>= 1, r >>= 1) {
            if (l & 1) res = max(res, t[l++]);
            if (!(r & 1)) res = max(res, t[r--]);
        }
        return res;
    }
};

int main(){
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n; cin >> n;
    vector<int> a(n), b;
    for (int i = 0; i < n; i++){
        cin >> a[i];
        b.push_back(a[i]);
    }
    sort(b.begin(), b.end());
    b.erase(unique(b.begin(), b.end()), b.end());
    int m = b.size();
    SegTree st(m);
    int best = 0;
    for (int x : a) {
        int idx = int(lower_bound(b.begin(), b.end(), x) - b.begin());
        int cur = 1;
        if (idx > 0) cur = st.query(0, idx-1) + 1;
        st.update(idx, cur);
        best = max(best, cur);
    }
    cout << best;
    return 0;
}
